src/odm_sharing/private/queries.py

```diff
@@ -1,14 +1,326 @@
+        table = n.str_val
+        select = n.sons[0]
+        filter_root = n.sons[1] if len(n.sons) > 1 else None
+        select_sql = f'SELECT {recurse(select)} FROM {ident(table)}'
+        filter_sql = f'WHERE {recurse(filter_root)}' if filter_root else ''
+        record(n, select_sql, [])
+        return (select_sql + ' ' + filter_sql)
+    elif n.kind == NodeKind.SELECT:
+        # select has an all-value or fields as children
+        if n.str_val == ALL_LIT:
+            return '*'
+        else:
+            columns = seq(n.sons).map(lambda x: ident(x.str_val))
+            return ','.join(columns)
+    elif n.kind == NodeKind.GROUP:
+        # group combines any number of children with its operator
+        op_str = n.str_val.upper()
+        assert op_str in ['AND', 'OR']
+        arg_start = len(args)
+        sql = seq(n.sons)\
+            .map(recurse)\
+            .reduce(lambda x, y: f'({x} {op_str} {y})')
+        record(n, sql, args[arg_start:])
+        return sql
+    elif n.kind == NodeKind.FILTER:
+        # filter has op as value, children define field, kind and literals
+        op = parse_op(n.str_val)
+        key_ident = ident(recurse(n.sons[0]))
+
+        def gen_range_sql(range_kind: RangeKind, values: List[str]) -> str:
+            '''generates sql for a range of values'''
+            if range_kind == RangeKind.INTERVAL:
+                a = values[0]
+                b = values[1]
+                return f'({key_ident} BETWEEN {a} AND {b})'
+            elif range_kind == RangeKind.SET:
+                return f"({key_ident} IN ({','.join(values)}))"
+            else:
+                assert False, "unreachable"
+                return ''
+
+        if op == Op.RANGE:
+            range_kind = RangeKind(n.sons[1].str_val)
+            assert range_kind != RangeKind.SINGLE
+            literals: List[str] = seq(n.sons[2:]).map(recurse).list()
+            sql = gen_range_sql(range_kind, literals)
+            record(n, sql, args[-len(literals):])
+            return sql
+        else:
+            val = recurse(n.sons[1])
+            sql = f'({key_ident} {op} {val})'
+            record(n, sql, args[-1:])
+            return sql
+    elif n.kind == NodeKind.FIELD:
+        # field name is quoted
+        return dqt(n.str_val)
+    elif n.kind == NodeKind.LITERAL:
+        # literal value is converted and added to list of args, while only a
+        # parameter placeholder is added to the sql
+        val = convert(n.str_val)
+        args.append(val)
+        return '?'
+    else:
+        assert False, 'unreachable'
+        return ''
+
+
+def gen_sql(
+    root_node: Node,
+    rule_queries: Dict[RuleId, PartialQuery],
+) -> Query:
+    '''generates sql from a node-tree
+
+    :param root_node: the root-node of the tree to generate sql from
+    :param rule_queries: (output) partial filter and select queries
+
+    :return: complete query
+    '''
+    args: List[str] = []
```

Comment:
it's an output-parameter for gen_data_sql (formely gen_sql_aux)

Reply:
Does it need to be declared here? Or can it be returned from the `gen_data_sql` method? Its confusing to see a variable declared and then never used.

Reply:
it's used since it's passed into the function on the next line. It's normal to do this for out-parameters, since the output-destination has to exist before the function is called. `gen_data_sql` is recursive, so it would be more difficult to implement it as a returned value.